yew/vlist: optimize diffing and patching

bakape opened the pull request:
#### Description

- Optimized keyed VList patching for node addition or removal from either end of the list and localized changes in the middle
- Changed unkeyed VList patching algorithm to much cheaper reverse vector itteration
- Added opt-in more verbose test logging
- Improved `Debug` output of some types

#### Checklist:

- [x] I have run `./ci/run_stable_checks.sh`
- [x] I have reviewed my own code
- [ ] I have added tests
<!-- If this is a bug fix, these tests will fail if the bug is present (to stop it from cropping up again) -->
<!-- If this is a feature, my tests prove that the feature works -->

<!-- Testing instructions -->
<!-- Check out the link below on how to setup and run tests -->
<!-- https://github.com/yewstack/yew/blob/master/CONTRIBUTING.md#test -->
<!-- If you're not sure how to test, let us know and we can provide guidance :) -->

<!-- Benchmark instructions -->
<!-- 1. Fork and clone https://github.com/yewstack/js-framework-benchmark -->
<!-- 2. Update `frameworks/yew/Cargo.toml` with your fork of Yew and the branch for this PR -->
<!-- 3. Open a new PR with your `Cargo.toml` changes -->
<!-- 4. Paste a link to the benchmark results: -->
- [x] I have opened a PR against https://github.com/yewstack/js-framework-benchmark: https://github.com/yewstack/js-framework-benchmark/pull/13

bakape commented:
yew-functional CI tests seem to have failed due to chromedriver.

jstarry commented:
@bakape perf looks good for `7b00d04e` but not for `0c8a2715`. Do you mind taking a look?

`7b00d04e`
```
result yew-05_swap1k.json min 49.149 max 52.446 mean 50.693000000000005 median 50.484 stddev 1.6584067655433616
result yew-baseline-05_swap1k.json min 49.25 max 52.606 mean 51.334 median 52.146 stddev 1.8193933054730091
```
`0c8a2715`
```
result yew-05_swap1k.json min 411.713 max 443.661 mean 429.0826666666667 median 431.874 stddev 16.155876093029836
result yew-baseline-05_swap1k.json min 51.105 max 60.054 mean 56.18866666666667 median 57.407 stddev 4.597216802950819
```

jstarry commented:
Also, I fixed the macro tests on master fyi

bakape commented:
The regression was introduced in e6c965b, which is very odd. I don't see anything there that could have caused it. I'll rebase and merge master anew.

bakape commented:
@jstarry Sorry, I've been busy with other matters and projects for a while. I've fixed the swap issue and added a few other optimizations. 

![Screenshot_20201119_033133](https://user-images.githubusercontent.com/7851952/99637834-169aee00-2a4e-11eb-82db-2f1ec302bda5.png)

jstarry commented:
No worries, I've been busy as well. Will review soon

bakape commented:
@siku2 
Merged master and reran benchmarks using bakape/js-framework-benchmark
![Screenshot_20210523_205748](https://user-images.githubusercontent.com/7851952/119271287-96766080-bc09-11eb-8ffb-4cbd53323e8a.png)

@jstarry 
yewstack/js-framework-benchmark is currently broken, so the action will fail. My PR should fix that.

cecton commented:
I'm looking forward to see this merged \o/

bakape commented:
It's been forgotten, it seems. A lot to review, so understandable finding time might be hard.

teymour-aldridge commented on packages/yew/src/utils.rs:
```diff
@@ -89,3 +89,13 @@ impl<IN, OUT> IntoIterator for NodeSeq<IN, OUT> {
     }
 }
+
+/// Print Node contents as a string for debugging purposes
```
nit

```suggestion
/// Print the `Node`'s contents as a string for debugging purposes
```

teymour-aldridge commented on packages/yew/src/virtual_dom/vlist.rs:
```diff
@@ -35,29 +74,235 @@ impl VList {
+    ) -> NodeRef {
+        use std::mem::{transmute, MaybeUninit};
+
+        macro_rules! map_keys {
```
I'm not sure about the benefit of using a macro here (especially with an eye towards compile times).

bakape replied:
It saves us from copy-pasting and I think the compile costs are negligible. Especially so, with how much the `html!` procedural macro does.

teymour-aldridge commented on packages/yew/src/virtual_dom/vnode.rs:
```diff
@@ -35,6 +35,16 @@ impl VNode {
         }
     }
 
+    /// Returns, if the VNode has a key without needlessly cloning the key
```
```suggestion
    /// Works out whether the `VNode` has a key (without cloning the key).
```

cecton replied:
Or maybe:
```suggestion
    /// Returns true if the `VNode` has a key without needlessly cloning the key.
```

teymour-aldridge replied:
Yeah either would work.

hamza1311 replied:
I'd do ```[`VNode`]``` instead so links to the type.

bakape replied:
This was done before I discovered doc links. I'll do that for the entire PR then.

teymour-aldridge reviewed (commented):
Just some thoughts

hamza1311 commented on packages/yew/src/virtual_dom/vtext.rs:
```diff
@@ -28,6 +28,20 @@ impl VText {
+            "VText {{ text: \"{}\", reference: {} }}",
+            self.text,
+            match &self.reference {
+                Some(_) => "Some(...)",
```
Nit: maybe also print the `VText`'s content? Or is that not possible?

bakape replied:
This was intentional, as it made the debug log output much more unreadable. Especially so in already verbose log output, like the one I added to the layout tests.

bakape replied:
This was intentional, as it made the debug log output much more unreadable. Especially so in already verbose log output, like the one I added to the layout tests.

hamza1311 commented on packages/yew/src/virtual_dom/mod.rs:
```diff
@@ -368,10 +368,10 @@ pub(crate) trait VDiff {
 }
 
-pub(crate) fn insert_node(node: &Node, parent: &Element, next_sibling: Option<Node>) {
+pub(crate) fn insert_node(node: &Node, parent: &Element, next_sibling: &Option<Node>) {
```
```suggestion
pub(crate) fn insert_node(node: &Node, parent: &Element, next_sibling: Option<&Node>) {
```
Taking `Option` itself as reference instead of the underlying data feels wrong

bakape replied:
Hmm, agreed.

hamza1311 reviewed (commented):
Apart from this comment, this looks good to me.
